internal/graph/check.go

```diff
@@ -38,6 +39,12 @@ type ResolveCheckRequest struct {
 	Context              *structpb.Struct
 	ResolutionMetadata   *ResolutionMetadata
 	VisitedPaths         map[string]struct{}
+	// This is different from ResolutionMetadata dispatch counter in that this counter needs to be
+	// incremented before dispatches are processed (where the ResolutionMetadata dispatch counter
+	// are incremented after the dispatches are processed). As this counter needs to be updated
+	// synchronously are multiple concurrent go-routine, it needs to use shared counter within
```

Comment:
> As this counter needs to be updated synchronously by multiple concurrent go-routine

This is not happening. Every time we call `dispatch` inside `LocalChecker`, the request object is a copy of the parent object, so you lose the "shared" aspect of it.